src/app/cluster/cluster-details/constraints/constraint-dialog/component.ts

```diff
@@ -0,0 +1,137 @@
+  constraint?: Constraint;
+}
+
+export enum Mode {
+  Add = 'add',
+  Edit = 'edit',
+}
+
+export enum Controls {
+  Name = 'name',
+  ConstraintTemplate = 'constraintTemplate',
+  Spec = 'spec',
+}
+
+@Component({
+  selector: 'km-constraint-dialog',
+  templateUrl: './template.html',
+  styleUrls: ['./style.scss'],
+})
+export class ConstraintDialog implements OnInit, OnDestroy {
+  readonly Controls = Controls;
+  form: FormGroup;
+  spec = '';
+  constraintTemplates: ConstraintTemplate[] = [];
+  private readonly _unsubscribe = new Subject<void>();
+
+  constructor(
+    public _matDialogRef: MatDialogRef<ConstraintDialog>,
+    private readonly _opaService: OPAService,
+    private readonly _notificationService: NotificationService,
+    @Inject(MAT_DIALOG_DATA) public data: ConstraintDialogConfig
+  ) {}
+
+  ngOnInit(): void {
+    this.form = new FormGroup({
+      [Controls.Name]: new FormControl(this.data.mode === Mode.Edit ? this.data.constraint.name : '', [
+        Validators.required,
+      ]),
+      [Controls.ConstraintTemplate]: new FormControl(
+        this.data.mode === Mode.Edit ? this.data.constraint.spec.constraintType.toLowerCase() : '',
+        [Validators.required]
+      ),
+    });
+
+    this._initConstraintSpecEditor();
+
+    this._opaService.constraintTemplates
+      .pipe(takeUntil(this._unsubscribe))
+      .subscribe(constraintTemplates => (this.constraintTemplates = constraintTemplates));
+  }
+
+  ngOnDestroy(): void {
+    this._unsubscribe.next();
+    this._unsubscribe.complete();
+  }
+
+  save(): void {
+    const constraint: Constraint = {
+      name: this.form.get(Controls.Name).value,
+      spec: this._getSpec(),
+    };
+
+    switch (this.data.mode) {
+      case Mode.Add:
+        return this._create(constraint);
+      case Mode.Edit:
+        return this._edit(constraint);
+    }
+  }
+
+  private _create(constraint: Constraint): void {
+    this._opaService
+      .createConstraint(this.data.projectId, this.data.clusterId, constraint)
+      .pipe(take(1))
+      .subscribe(result => {
+        this._matDialogRef.close(true);
+        this._notificationService.success(`The constraint ${result.name} was created`);
+        this._opaService.refreshConstraint();
+      });
+  }
+
+  private _edit(constraint: Constraint): void {
+    this._opaService
+      .patchConstraint(this.data.projectId, this.data.clusterId, this.data.constraint.name, constraint)
+      .pipe(take(1))
+      .subscribe(result => {
+        this._matDialogRef.close(true);
+        this._notificationService.success(`The constraint ${result.name} was updated`);
+        this._opaService.refreshConstraint();
+      });
+  }
+
+  private _initConstraintSpecEditor(): void {
+    if (this.data.mode === Mode.Edit) {
+      const spec = this.data.constraint.spec;
+      if (!_.isEmpty(spec)) {
+        this.spec = dump(spec);
+      }
+    }
+  }
+
+  private _getSpec(): any {
```

Comment:
This method is quite confusing. I'd add here at least some basic types to better identify the loaded object. We should be able to expect some kind of object schema loaded from the spec and cast it to a typed object instead of any.

Also, what is even more confusing here is that the object is loaded, then there is no check but something is assigned to a field of this `any` object and then it is being checked if it is empty. Is there an actual reason for this?

Reply:
Yeah I also don't like it. I tried a bit to improve it in https://github.com/kubermatic/dashboard/pull/3008/commits/60bfb67a8f077edad78a0cb4ce3df7f804f39073, but I'm not sure if we can do it better.... Any input is welcome.

Also I'm wondering if we should use the `Constraint Template` from the form that we have or from the spec (if specified). I guess I think it is saver to use the form, as this field only displays existing templates. 🤔

Reply:
@floreks ping ;)